refactored huge switch; fixed null id. in dev mode

machisuji opened the pull request.

machisuji committed:
refactored huge switch; fixed null id. in dev mode

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://machisuji-fix-api-null-identifier.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-fix-api-null-identifier.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
:warning: We need to merge this into `release/4.2` :warning: 

@machisuji Did you already base this branch on `release/4.2`? Or is it based on the `dev`?

Since the `release/4.2` branch is already open, we need to target all our work against that branch and only target stuff against `dev` that we do not want in too quickly ^^

NobodysNightmare commented:
Specs failing

myabc commented on lib/api/v3/errors/error_representer.rb:
```diff
@@ -57,30 +57,7 @@ def _type
-          when ::API::Errors::UnsupportedMediaType
-            'urn:openproject-org:api:v3:errors:TypeNotSupported'
-          end
+          represented.class.identifier rescue nil
```
``` ruby
 represented.class.identifier if represented.class.respond_to?(:identifier)
```

NobodysNightmare replied:
@myabc :+1:

ulferts replied:
The change looks good codewise. However there is one drawback. It makes the formerly not version specific api errors version specific because of the namespace e.g. `urn:openproject-org:api:v3:errors:UpdateConflict`. 

We should either decide that the errors are api version specific and then move their code into the API::V3 errors namespace. Or we decide that they are not and find a different way to refactor this. My beloved way of hiding case statements by introducing a hash comes to mind.

NobodysNightmare replied:
I'd rather move those errors into APIv3 then:
- APIv4 or whatever you'd like to call it is not on the horizon, not even in your wildest dreams I hope
- I doubt that those errors will provide any meaningful help to any other version of an API and I'd rather not plan ahead for such uncertainties

So I'd vote to move the errors into the V3 namespace...

floriank commented:
@machisuji I know you're busy, but ... ping?

machisuji commented:
@floriank Yeah sorry I have yet to create my PR for the WWW-Authenticate header business. Should hopefully happen today. I'm on it. Dunno when I'll get to this switch business, though. If we still want this perhaps someone else has to take over to fix it as I am quite busy currently.

NobodysNightmare commented:
@machisuji The problem is that now there is a release soon and this makes some bug fixes urgent...

I will have a look at this piece of art. May I get collab access to your fork of OP?

machisuji commented:
Sure.

NobodysNightmare committed:
do not use exceptions for code flow

I have to admit that I personally would not even care about this error condition, as I'd expect my errors to listen to :identifier

NobodysNightmare committed:
specify identifier for parse error

NobodysNightmare commented on lib/api/errors/parse_error.rb:
```diff
@@ -30,6 +30,8 @@
 module API
   module Errors
     class ParseError < InvalidRequestBody
+      identifier InvalidRequestBody.identifier
```
This one is quite tricky. Inheritance does not work for identifiers.

I chose the lazy approach of simply declaring to use the same identifier as the parent class. The alternative would be to change `identifier` so that it implicitly "inherits" from the superclass, however I think the explicit approach might be more sensible here.

NobodysNightmare commented:
That last one was actually a real `null` identifier ^^

I'd call this done if travis agrees. However, someone else might want to have a look at my last two commits, otherwise I would merge my own stuff without review ^^

NobodysNightmare commented:
Tests are flickering...

Azure7111 commented:
LGTM :+1: 

Does anybody know why the coverage is so low? This seems like a false report in my opinion.

Anyway - Will merge when green.
